spring-integration-core/src/main/java/org/springframework/integration/support/AcknowledgmentCallbackFactory.java

```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.integration.support;
+
+import java.util.function.Supplier;
+
+/**
+ * A factory for creating {@link AcknowledgmentCallback}s.
+ *
+ * @param <T> a type containing information with which to populate the acknowledgment.
+ *
+ * @author Gary Russell
+ * @since 5.0.1
+ *
+ */
+public interface AcknowledgmentCallbackFactory<T> {
+
+	/**
+	 * Create the callback.
+	 * @param supplier a supplier of T
+	 * @return the callback
+	 */
+	AcknowledgmentCallback createCallback(Supplier<T> supplier);
```

Comment:
I wonder why would one need to customize such a low level resource...
Just curious what is the point of this abstraction.
And why have you made it based on the `Supplier`... Why simple `T` is not enough?..

Thanks

Reply:
I thought about keeping it simple but decided that using a factory would enable users to customize the ack behavior. For example, with Kafka (or Rabbit), the reject could be modified to send the message to another topic.

I think these extension points are useful.

Reply:
But I agree, the Supplier is not needed.

Reply:
OK with factory, but not OK with `Supplier` 😄